renku/service/jobs/graph.py

```diff
@@ -0,0 +1,127 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Renku graph jobs."""
+import tempfile
+from urllib.parse import urlparse
+
+from git import GitError, Repo
+from marshmallow import EXCLUDE
+from requests import RequestException
+from sentry_sdk import capture_exception
+
+from renku.core.commands.format.graph import jsonld
+from renku.core.commands.graph import build_graph
+from renku.core.commands.migrate import migrate_project
+from renku.core.errors import MigrationError, RenkuException
+from renku.core.utils.contexts import chdir
+from renku.core.utils.requests import retry
+from renku.service.errors import RenkuOpTimeoutError
+from renku.service.serializers.cache import ProjectCloneContext
+from renku.service.serializers.graph import GraphBuildCallbackError, GraphBuildCallbackSuccess
+from renku.service.utils.timeout import timeout
+
+
+def report_recoverable(payload, exception, callback_url):
+    """Report to callback URL recoverable state."""
+    capture_exception(exception)
+
+    if not callback_url:
+        return
+
+    payload["failure"] = {"type": "RECOVERABLE_FAILURE", "message": str(exception)}
+
+    data = GraphBuildCallbackError().load(payload)
+    with retry() as session:
+        session.post(callback_url, data=data)
+
+
+def report_unrecoverable(payload, exception, callback_url):
+    """Report to callback URL unrecoverable state."""
+    capture_exception(exception)
+
+    if not callback_url:
+        return
+
+    payload["failure"] = {"type": "UNRECOVERABLE_FAILURE", "message": str(exception)}
+
+    data = GraphBuildCallbackError().load(payload)
+    with retry() as session:
+        session.post(callback_url, data=data)
+
+
+def report_success(request_payload, graph_payload, callback_url):
+    """Report to callback URL success state."""
+    data = GraphBuildCallbackSuccess().load({**request_payload, **graph_payload})
+
+    if not callback_url:
+        return data
+
+    with retry() as session:
+        session.post(callback_url, data=data)
+
+    return data
+
+
+def _build_and_report(callback_payload, callback_url, ctx):
+    """Build graph and report on result."""
+    with tempfile.TemporaryDirectory() as tmpdir:
+        try:
+            Repo.clone_from(ctx["url_with_auth"], tmpdir)
+        except GitError as e:
+            report_recoverable(callback_payload, e, callback_url)
+
+        with chdir(tmpdir):
+            try:
+                command = migrate_project().with_commit().build()
+
+                result = command.execute(skip_template_update=True, skip_docker_update=True)
+                result, _, _ = result.output
+
+                if result:
+                    graph = build_graph(revision=callback_payload["commit_id"])
```

Comment:
Same as above about revision. We'd need to checkout `revision` instead, run migrate on top of that, then call `build_graph` without revision.

Reply:
Done. :rocket: